crates/store/src/db/sql.rs

```diff
@@ -1042,19 +945,14 @@ pub fn get_state_sync(
         block_num,
     )?;
 
-    let (block_header, chain_tip) = if !notes.is_empty() {
-        let block_header = select_block_header_by_block_num(conn, Some(notes[0].block_num))?
-            .ok_or(StateSyncError::EmptyBlockHeadersTable)?;
-        let tip = select_block_header_by_block_num(conn, None)?
-            .ok_or(StateSyncError::EmptyBlockHeadersTable)?;
-
-        (block_header, tip.block_num())
-    } else {
-        let block_header = select_block_header_by_block_num(conn, None)?
-            .ok_or(StateSyncError::EmptyBlockHeadersTable)?;
-
-        let block_num = block_header.block_num();
-        (block_header, block_num)
+    let tip = select_block_header_by_block_num(conn, None)?
+        .ok_or(StateSyncError::EmptyBlockHeadersTable)?;
+    let chain_tip = tip.block_num();
+    let block_header = match notes.first() {
+        None => tip,
+        Some(note) if note.block_num == chain_tip => tip,
+        Some(note) => select_block_header_by_block_num(conn, Some(note.block_num))?
+            .ok_or(StateSyncError::EmptyBlockHeadersTable)?,
```

Comment:
iiuc this could be simplified:

```suggestion
    let block_header = notes.first().map(|note| note.block_num);
    let block_header = select_block_header_by_block_num(conn, block_header)?
        // This error variant should likely be missing block header {n}.
        .ok_or(StateSyncError::EmptyBlockHeadersTable)?;
```

Reply:
Thank you for the suggestion, but we need to return a chain tip from the method, thus we still need two `select_block_header_by_block_num` requests. In my solution we reuse information from the first request only in a case when block header points to the latest block, otherwise, we do second request.

Reply:
Ah right apologies. The github diff view had the return cutoff and I didn't think to look ahead.